honeybadgermpc/betterpairing.py

```diff
@@ -312,18 +335,127 @@ def rand(seed=None):
         return G2(out)
 
 
+class GT:
+    def __init__(self, other=None):
+        if other is None:
+            self.pyfq12 = PyFq12()
+            self.pyfq12.rand(1, 0, 0, 0)
+        elif type(other) is PyFq12:
+            self.pyfq12 = other
+        elif type(other) is list:
+            assert len(other) == 12
+            self.pyfq12 = PyFq12()
+            self.pyfq12.from_strs(*other)
+        elif type(other) is int:
+            self.pyfq12 = PyFq12()
+            self.pyfq12.from_strs(str(other), *["0"]*11)
+        elif type(other) is str:
+            l = [x.strip() for x in other.split(',')]
+            assert len(l) == 12
+            if l[0][1] == 'x':
+                for i in range(len(l)):
+                    l[i] = str(int(l[i]))
+            self.pyfq12 = PyFq12()
+            self.pyfq12.from_strs(*l)
+
+    def __mul__(self, other):
+        if type(other) is int:
+            out = GT(dupe_pyfq12(self.pyfq12))
+            if other == 0:
+                out.pyfq12.zero()
+                return out
+            if other < 0:
+                out.pyfq12.negate()
+                other = -1 * other
+            prodend = GT(other)
+            out.pyfq12.mul_assign(prodend.pyfq12)
+            return out
+        elif type(other) is ZR:
+            out = GT(dupe_pyfq12(self.pyfq12))
+            out.pyfq12.mul_assign(other.val)
+            return out
+        elif type(other) is GT:
+            out = GT(dupe_pyfq12(self.pyfq12))
+            out.pyfq12.mul_assign(other.pyfq12)
+            return out
+        else:
+            raise TypeError(
+                'Invalid exponentiation param. Expected ZR or int. Got ' + str(type(other)))
+
+    def __add__(self, other):
+        if type(other) is GT:
+            out = dupe_pyfq12(self.pyfq12)
+            out.add_assign(other.pyfq12)
+            return G2(out)
+
+    # TODO: __truediv__ needs to be implemented
+
+    # TODO: __pow__ needs to be implemented
```

Comment:
``__truediv__`` and ``__pow__`` are not implemented. They are called below.

Reply:
I think we should narrow this interface abstraction a bit. G1 G2 and GT are both groups, but G1 and G2 are additive notation (you can add two points together, but not multiply them together).

   Now the slightly confusing thing is that GT is actually defined as the finite field Fq12.... and in general finite fields support multiplication as well as addition. However, the GT class should probably not expose `__add__` at all.